Fixed #23895 -- Prevented pickling of ResolverMatch.

jonathandavis4 opened the pull request:
Fixes issue [#23895](https://code.djangoproject.com/ticket/23895), where attempting to pickle a `ResolverMatch` for a class-based view causes an exception.

auvipy reviewed (changes requested):
please tun isort and flake8 to your pr branch

felixxm commented on django/urls/resolvers.py:
```diff
@@ -71,6 +72,27 @@ def __repr__(self):
+        if self.is_class_based:
+            # The following attributes of ResolverMatch prevent pickling.
+            del state['func']
+            del state['tried']
```
`tried` can be also an issue for function-based views because `URLResolver` is not pickleable, everything depends on the order in `urlpatterns`, see e.g.
```diff
diff --git a/tests/urlpatterns/path_urls.py b/tests/urlpatterns/path_urls.py
index 29b6185292..0efb713959 100644
--- a/tests/urlpatterns/path_urls.py
+++ b/tests/urlpatterns/path_urls.py
@@ -3,13 +3,13 @@ from django.urls import include, path, re_path
 from . import views
 
 urlpatterns = [
+    path('included_urls/', include('urlpatterns.included_urls')),
     path('articles/2003/', views.empty_view, name='articles-2003'),
     path('articles/<int:year>/', views.empty_view, name='articles-year'),
     path('articles/<int:year>/<int:month>/', views.empty_view, name='articles-year-month'),
     path('articles/<int:year>/<int:month>/<int:day>/', views.empty_view, name='articles-year-month-day'),
     path('users/', views.empty_view, name='users'),
     path('users/<id>/', views.empty_view, name='user-with-id'),
-    path('included_urls/', include('urlpatterns.included_urls')),
     re_path(r'^regex/(?P<pk>[0-9]+)/$', views.empty_view, name='regex'),
     re_path(r'^regex_optional/(?P<arg1>\d+)/(?:(?P<arg2>\d+)/)?', views.empty_view, name='regex_optional'),
     re_path(
```

felixxm commented on django/urls/resolvers.py:
```diff
@@ -71,6 +72,27 @@ def __repr__(self):
+        self.__dict__.update(state)
+
+        if self.is_class_based:
+            # Set the attributes again which could not be pickled.
+            path = '/{}'.format(self.route)
+            match = resolve(path)
+            self.func = match.func
+            self.tried = match.tried
```
URL patterns may change over time, so we can restore the different match :thinking:

felixxm reviewed (commented):
@jonathandavis4 Thanks for this patch :+1: Changing `__getstate__()`/`__setstate__()` is always tricky, unfortunately I have no idea for a better alternative. 

@kezabelle Do you want to take a look?

kezabelle commented:
So, it's been a heck of a long time since I actively thought about this, and the frequency with which I need to pickle anything means I'm always muddy on the details (as @felixxm said, reduce/getstate/setstate are always tricky ;)).

I can't actually _recall_ what's _in_ `.tried` that directly precludes it from being pickled, and unfortunately the tests aren't illuminating me. Whilst I don't doubt that it's also problematic, can someone give me an example? It may be possible for whatever is in there to become pickleable of it's own accord and not have to special-case it here, perhaps? 

Conceptually, at a basic level I _think_ it ought to be possible to replace `.func` with an instance of the class when serialising, and re-inflate it to a callable when deserialized. I may be missing the nuances though, so forgive me if you already established it wasn't workable.
ie: a CBV which correctly inherits from View intrinsically has `as_view()` and as a result has `.view_class` and `.view_initkwargs` both of which are _generally speaking_ individually pickleable (ignoring the possibility that the user actively puts a non-pickleable item into the initkwargs, which is sort of on them)

So it _may_ be possible to avoid introducing `self.is_class_based` and have something like, in the getstate/reduce portion:
```
...
if hasattr(state.func, 'view_class') and hasattr(state.func, 'view_initkwargs'):
    state.func= state.func.view_class(**state.func.view_initkwargs)
else:
    ...
```
and then in the deserializing portion:
```
....
if hasattr(state.func, 'as_view'):
    state.func = state.func.as_view()
...
```

I'll admit that I've not tried it, and there may be other reasons that doesn't work :) It '_solves_' the issue using `resolve()` would present where it could theoretically now point to a different callable, by freezing the reference in time, but there's a couple of other things that would need further investigation in either scenario I guess:
- does the proposed solution, or the possible alternative outlined above, behave well with any method/class decorating? Gut feeling is _yes_, but I'm up for being proven wrong...
- what happens if the resolver match callable was decorated, (eg the urlpattern definition was: `login_required(myview)` or `login_required(MyCustomView.as_view())`) ... can those be serialized, or if they can't, the wrappers be unwrapped enough to actually facilitate reconstruction in setstate? 
If any of those kinds of usages present a problem for CBVs _or_ FBVs (or both), I'd be inclined to to make _all_ ResolverMatch throw an exception on serialization through `__reduce_ex__` or whatever, so the downstream user explicitly knows they have to handle it rather than it bite them on only some callables. If serialization generally works for those for both, barring initkwargs themselves not being serializable because of the user, I think that'd be fine.

Was that coherent? Do shout if not :)

felixxm commented:
> Interesting. I think `URLPattern` _could_ be separately serialisable in a similar way -- I'm presuming serialising the callback is the intrinsic problem, as it is here WRT CBVs & FBVs.

`URLPattern` is pickleable, but `.tried` can contain also `URLResolver` (see https://github.com/django/django/pull/14664#discussion_r675332227) which is not pickleable because it contains a `module`.

kezabelle commented:
> `URLPattern` is pickleable, but `.tried` can contain also `URLResolver` which is not pickleable because it contains a `module`.

Ah, well that's rather intractable. Alas that it can be the case instead of being unrolled!
I suppose one _could_ reconstruct the CBV as I suggested above (syntax notwithstanding, I can plainly see it's wrong now I look at it again), and use `resolve()` to get _a_ list of tried items (if not _the_ list) on the basis that then you'd have The Correct Callable and it's ostensibly rare that you'd want to examine the `tried` list so it might not matter that it _could_ be wrong. But then we'd still have to consider whether or not it's _really_ the correct callable if you don't go through `resolve` as it might not have serialized wrappers/decorators (at first glance, anyway) ... and going through `resolve` could _also_ not be the The Correct Callable.

If it's not generally possible to construct the correct state for pickling, and we don't want to use `resolve()` for the reasons you mentioned upthread (and/or because it also involves going through the threadlocal perhaps undesirably) then maybe it's best just to prevent any calls to it. 
It's got to be pretty infrequently attempted (given how long the ticket has been open and the lack of marked duplicates) so maybe just an anti-footgun in the form of an informative exception for anyone who tries it is enough? The other option could well be that Jonathan (or someone else in the future) can get something that's 80/20% workable and it still not be satisfactory, which would be less than ideal too from both a time investment (for them) and maintenance (for you) perspective.

jonathandavis4 commented:
Hi @felixxm and @kezabelle, thanks for your comments!

To summarise and confirm my understanding is correct:

1. Using `resolve` to rebuild the `ResolverMatch` is not enough, as the currently active URL conf may have changed since the object was pickled.
    1.1 We cannot check whether `URLResolver.urlconf_name` has changed to work out whether `resolve` would return the correct callable for the same reason, as the same module could have different URLs configured.
2. With regards to `func`, we might be able to replace it with a class instance when pickling and re-inflate it when unpickling.
3. With regards to `tried`:
    3.1. It cannot be pickled as it contains module objects.
    3.2. We would need to handle attempts to pickle it for function-based views too.
4. We would need to ensure the solution works nicely with decorated views too, possibly by unwrapping and reconstructing them if needed.
5. The alternative to all of the above is to raise an exception with an explanatory message when `ResolverMatch.__setstate__` is called.

I was wondering if we might want to go ahead with 1-4, but mention in the URLs documentation somewhere that it is only guaranteed to be possible to unpickle `ResolverMatch` if the URL conf hasn't changed since it was pickled? (It would be nice if we could pickle something like a URL conf version number along with it and raise an Exception or at least log a warning if one is unpickled with a different version, but I'm not sure any information like that is available?)

It would mean that for this PR, we can use `resolve` in the implementation as it currently is, which is likely cleaner than pickling a class instance for `func` (though probably slower), and gets around the problem with pickling modules in `tried`.

For the user, `__setstate__` and `__getstate__` will still be available. In cases where the URL confs are changing, it is likely they will need to add extra functionality in their applications to handle this.

If we did this, I would just need to add tests for checking it works correctly with decorated views, and update the documentation.

Let me know if I'm misunderstanding something. I'm happy to go with whichever solution you decide is best. :)

felixxm commented:
I don't think that we need/should support pickling `ResolverMatch` so I'd raise an exception or even leave it as it is.

@tomchristie's [comment](https://code.djangoproject.com/ticket/23895#comment:2) about pickling `HttpResponse` is IMO more important and we should focus on fixing it:
```
>>> import pickle
>>> response = self.client.get('/cbv_view/')
>>> pickle.dumps(response)
...
AttributeError: Can't pickle ....
```

kezabelle commented:
> raise an exception or even leave it as it is.

I'd prefer the exception, because 7 years ago it was a confusing annoyance ;)

Something like:
```
def __reduce_ex__(self, protocol):
    raise PicklingError(f"Can't pickle {type(self)}")
```
would probably do, non? It was the inconsistency between CBVs and FBVs that tripped _me_, rather than the general pickleability of the thing. IIRC I just ended up doing a `del` on the attr once I figured out what was going on.

Agree that Tom's use-case is something that _should_ be fixed if possible, which would almost certainly need to use getstate/setstate shenanigans and be more selective about what gets persisted, even at the loss of some fidelity WRT to attributes getting frozen & restored. Some of the same sticking points may well exist (eg: the `resolver_match`, the `urlconf` getting set by a middleware, etc) of course and may need to be ignored for the purpose of getting just a `fresh HttpResponse with correct content, status code and headers to be reconstructed`

felixxm commented:
> I'd prefer the exception, because 7 years ago it was a confusing annoyance ;)
> 
> Something like:
> 
> ```
> def __reduce_ex__(self, protocol):
>     raise PicklingError(f"Can't pickle {type(self)}")
> ```

Agreed, let's raise an exception and then we can focus this ticket (and the 2nd PR) on fixing pickleability for `HttpResponse`.

@jonathandavis4 Do you have time to keep working on this?

jonathandavis4 commented:
@felixxm Sure, I can take a look at raising the exception in this PR tonight, then I can start a new PR for picking `HttpResponse`s over the weekend maybe.

jonathandavis4 commented:
@felixxm @kezabelle I have now made the change to raise an exception when pickling a `ResolverMatch`, updated the release notes and squashed the commits. Please let me know if anything needs changing. Thanks.

felixxm commented on tests/urlpatterns/test_pickling.py:
```diff
@@ -0,0 +1,20 @@
+        """Test that attempting to pickle a ResolverMatch causes a
+        PicklingError to be raised.
+        """
+        match = resolve('/users/')
+
+        self.assertIsInstance(match, ResolverMatch)
+
+        with self.assertRaises(PicklingError):
+            dumps(match)
```
```suggestion
        msg = 'Cannot pickle ResolverMatch.'
        with self.assertRaisesMessage(pickle.PicklingError, msg):
            pickle.dumps(resolve('/users/'))
```

felixxm reviewed (commented):
@jonathandavis4 Thanks :+1: I left comments.

felixxm commented:
> Agreed, let's raise an exception and then we can focus this ticket (and the 2nd PR) on fixing pickleability for `HttpResponse`.

Pickleability for `HttpResponse` deserves a separate ticket.

jonathandavis4 commented:
@felixxm Those changes have been made now, and I have created a [new ticket](https://code.djangoproject.com/ticket/32969) for pickling `HTTPResponse`s. :)

felixxm commented:
@jonathandavis4 Thanks :+1: Welcome aboard :boat:

tigicion commented:
@felixxm @jonathandavis4 I have created a pull request on fixing pickleability for HttpResponse. Please let me know if this is the right solution.
[https://github.com/django/django/pull/14728](url)
